References
----------

.. [Amid2019] Amid & Warmuth (2019),
   *TriMap: Large-scale Dimensionality Reduction Using Triplets*,
   `arXiv <https://arxiv.org/abs/1910.00204>`__.

.. [Amir2013] Amir *et al.* (2013),
   *viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.2594>`__.

.. [Angerer2016] Angerer *et al.* (2016),
   *destiny â€“ diffusion maps for large-scale single-cell data in R*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv715>`__.

.. [Bernstein2020] Bernstein *et al.* (2020),
  *Solo: Doublet Identification in Single-Cell RNA-Seq via Semi-Supervised Deep Learning*,
  `Cell Systems  <https://doi.org/10.1016/j.cels.2020.05.010>`__.

.. [Blondel2008] Blondel *et al.* (2008),
   *Fast unfolding of communities in large networks*,
   `J. Stat. Mech. <https://doi.org/10.1088/1742-5468/2008/10/P10008>`__.

.. [Chippada2018]
   *ForceAtlas2 for Python and NetworkX*,
   `GitHub <https://github.com/bhargavchippada/forceatlas2>`__.

.. [Coifman2005] Coifman *et al.* (2005),
   *Geometric diffusions as a tool for harmonic analysis and structure definition of data: Diffusion maps*,
   `PNAS <https://doi.org/10.1073/pnas.0500334102>`__.

.. [Csardi2006] Csardi *et al.* (2006),
   *The igraph software package for complex network research*,
   `InterJournal Complex Systems <https://igraph.org/>`__.

.. [Eraslan2018] Eraslan and Simon *et al.* (2018),
   *Single cell RNA-seq denoising using a deep count autoencoder*,
   `bioRxiv <https://doi.org/10.1101/300681>`__.

.. [Fechtner2018]
   *PyPairs*,
   `GitHub <https://github.com/rfechtner/pypairs>`__.

.. [Fruchterman2091] Fruchterman & Reingold (1991),
   *Graph drawing by force-directed placement*,
   `Software: Practice & Experience <https://doi.org/10.1002/spe.4380211102>`__.

.. [Gardner2000] Gardner *et al.*, (2000)
   *Construction of a genetic toggle switch in Escherichia coli*,
   `Nature <https://doi.org/10.1038/35002131>`__.

.. [Haghverdi2015] Haghverdi *et al.* (2015),
   *Diffusion maps for high-dimensional single-cell analysis of differentiation data*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv325>`__.

.. [Haghverdi2016] Haghverdi *et al.* (2016),
   *Diffusion pseudotime robustly reconstructs branching cellular lineages*,
   `Nature Methods <https://doi.org/10.1038/nmeth.3971>`__.

.. [Haghverdi2018] Haghverdi *et al.* (2018),
   *Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.4091>`__.

.. [Hie2019] Hie *et al.* (2019),
   *Efficient integration of heterogeneous single-cell transcriptomes using Scanorama*,
   `Nature Biotechnology <https://doi.org/10.1038/s41587-019-0113-3>`__.

.. [Islam2011] Islam *et al.* (2011),
   *Characterization of the single-cell transcriptional landscape by highly multiplex RNA-seq*,
   `Genome Research <https://doi.org/10.1101/gr.110882.110>`__.

.. [Jacomy2014] Jacomy *et al.* (2014),
   *ForceAtlas2, a Continuous Graph Layout Algorithm for Handy Network Visualization Designed for the Gephi Software*
   `PLOS One <https://doi.org/10.1371/journal.pone.0098679>`__.

.. [Johnson2007] Johnson, Li & Rabinovic (2007),
   *Adjusting batch effects in microarray expression data using empirical Bayes methods*,
   `Biostatistics <https://doi.org/10.1093/biostatistics/kxj037>`__.

.. [Kang2018] Kang *et al.* (2018),
   *Python Implementation of MNN correct*,
   `GitHub <https://github.com/chriscainx/mnnpy>`__.

.. [Korunsky2019] Korunsky *et al.* (2019),
   *Fast, sensitive and accurate integration of single-cell data with Harmony*,
   `Nature Methods <https://doi.org/10.1038/s41592-019-0619-0>` __.

.. [Krumsiek2011] Krumsiek *et al.* (2011),
   *Hierarchical Differentiation of Myeloid Progenitors Is Encoded in the Transcription Factor Network*,
   `PLoS ONE <https://doi.org/10.1371/journal.pone.0022649>`__.

.. [Lambiotte2009] Lambiotte *et al.* (2009)
   *Laplacian Dynamics and Multiscale Modular Structure in Networks*
   `arXiv <https://arxiv.org/abs/0812.1770>`__.

.. [Lause2021] Lause *et al.* (2021)
   *Analytic Pearson residuals for normalization of single-cell RNA-seq UMI data*,
   `Genome Biology <https://doi.org/10.1186/s13059-021-02451-7>`__.

.. [Leek2012] Leek *et al.* (2012),
   *sva: Surrogate Variable Analysis. R package*
   `Bioconductor <https://doi.org/10.18129/B9.bioc.sva>`__.

.. [Levine2015] Levine *et al.* (2015),
   *Data-Driven Phenotypic Dissection of AML Reveals Progenitor--like Cells that Correlate with Prognosis*,
   `Cell <https://doi.org/10.1016/j.cell.2015.05.047>`__.

.. [Maaten2008] Maaten & Hinton (2008),
   *Visualizing data using t-SNE*,
   `JMLR <https://www.jmlr.org/papers/v9/vandermaaten08a.html>`__.

.. [McCarthy2017] McCarthy *et al.* (2017),
   *scater: Single-cell analysis toolkit for gene expression data in R*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btw777>`__.

.. [Moon2017] Moon *et al.* (2019),
   *PHATE: A Dimensionality Reduction Method for Visualizing Trajectory Structures in High-Dimensional Biological Data*,
   `Nature Biotechnology <https://doi.org/10.1038/s41587-019-0336-3>`__.

.. [Satija2015] Satija *et al.* (2015),
   *Spatial reconstruction of single-cell gene expression data*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.3192>`__.

.. [Manno2018] La Manno *et al.* (2018),
   *RNA velocity of single cells*,
   `Nature <https://doi.org/10.1038/s41586-018-0414-6>`__.

.. [McInnes2018] McInnes & Healy (2018),
   *UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction*,
   `arXiv <https://arxiv.org/abs/1802.03426>`__.

.. [Moignard2015] Moignard *et al.* (2015),
   *Decoding the regulatory network of early blood development from single-cell gene expression measurements*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.3154>`__.

.. [Ntranos2018] Ntranos *et al.* (2018),
   *Identification of transcriptional signatures for cell types from single-cell RNA-Seq*,
   `bioRxiv <https://doi.org/10.1101/258566>`__.

.. [Paul2015] Paul *et al.* (2015),
   *Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors*,
   `Cell <https://doi.org/10.1016/j.cell.2015.11.013>`__.

.. bibliography::
