References
----------

.. [Amid19] Amid & Warmuth (2019),
   *TriMap: Large-scale Dimensionality Reduction Using Triplets*,
   `arXiv <https://arxiv.org/abs/1910.00204>`__.

.. [Amir13] Amir *et al.* (2013),
   *viSNE enables visualization of high dimensional single-cell data and reveals phenotypic heterogeneity of leukemia*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.2594>`__.

.. [Angerer16] Angerer *et al.* (2016),
   *destiny – diffusion maps for large-scale single-cell data in R*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv715>`__.

.. [Bernstein20] Bernstein *et al.* (2020),
  *Solo: Doublet Identification in Single-Cell RNA-Seq via Semi-Supervised Deep Learning*,
  `Cell Systems  <https://doi.org/10.1016/j.cels.2020.05.010>`__.

.. [Blondel08] Blondel *et al.* (2008),
   *Fast unfolding of communities in large networks*,
   `J. Stat. Mech. <https://doi.org/10.1088/1742-5468/2008/10/P10008>`__.

.. [Chippada18]
   *ForceAtlas2 for Python and NetworkX*,
   `GitHub <https://github.com/bhargavchippada/forceatlas2>`__.

.. [Coifman05] Coifman *et al.* (2005),
   *Geometric diffusions as a tool for harmonic analysis and structure definition of data: Diffusion maps*,
   `PNAS <https://doi.org/10.1038/nmeth.3971>`__.

.. [Csardi06] Csardi *et al.* (2006),
   *The igraph software package for complex network research*,
   `InterJournal Complex Systems <http://igraph.org>`__.

.. [Eraslan18] Eraslan and Simon *et al.* (2018),
   *Single cell RNA-seq denoising using a deep count autoencoder*,
   `bioRxiv <https://doi.org/10.1101/300681>`__.

.. [Ester96] Ester *et al.* (1996),
   *A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases with Noise*,
   `Proceedings of the 2nd International Conference on Knowledge Discovery and Data Mining,
   Portland, OR <http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.121.9220>`__.

.. [Eulenberg17] Eulenberg *et al.* (2017),
   *Reconstructing cell cycle and disease progression using deep learning*
   `Nature Communications <https://doi.org/10.1101/081364>`__.

.. [Fechtner18]
   *PyPairs*,
   `GitHub <https://github.com/rfechtner/pypairs>`__.

.. [Fruchterman91] Fruchterman & Reingold (1991),
   *Graph drawing by force-directed placement*,
   `Software: Practice & Experience <http://doi.org:10.1002/spe.4380211102>`__.

.. [Gardner00] Gardner *et al.*, (2000)
   *Construction of a genetic toggle switch in Escherichia coli*,
   `Nature <https://doi.org/10.1038/35002131>`__.

.. [Hagberg08] Hagberg *et al.* (2008),
   *Exploring Network Structure, Dynamics, and Function using NetworkX*,
   `Scipy Conference <http://conference.scipy.org/proceedings/SciPy2008/paper_2/>`__.

.. [Hastie09]
   Hastie *et al.* (2009),
   *The Elements of Statistical Learning*,
   `Springer <https://web.stanford.edu/~hastie/ElemStatLearn/>`__.

.. [Haghverdi15] Haghverdi *et al.* (2015),
   *Diffusion maps for high-dimensional single-cell analysis of differentiation data*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv325>`__.

.. [Haghverdi16] Haghverdi *et al.* (2016),
   *Diffusion pseudotime robustly reconstructs branching cellular lineages*,
   `Nature Methods <https://doi.org/10.1038/nmeth.3971>`__.

.. [Haghverdi18] Haghverdi *et al.* (2018),
   *Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.4091>`__.

.. [Hie19] Hie *et al.* (2019),
   *Efficient integration of heterogeneous single-cell transcriptomes using Scanorama*,
   `Nature Biotechnology <https://doi.org/10.1038/s41587-019-0113-3>`__.

.. [Huber15] Huber *et al.* (2015),
   *Orchestrating high-throughput genomic analysis with Bioconductor*,
   `Nature Methods <https://doi.org/10.1038/nmeth.3252>`__.

.. [Islam11] Islam *et al.* (2011),
   *Characterization of the single-cell transcriptional landscape by highly multiplex RNA-seq*,
   `Genome Research <https://doi.org/10.1101/gr.110882.110>`__.

.. [Jacomy14] Jacomy *et al.* (2014),
   *ForceAtlas2, a Continuous Graph Layout Algorithm for Handy Network Visualization Designed for the Gephi Software*
   `PLOS One <https://doi.org/10.1371/journal.pone.0098679>`__.

.. [Johnson07] Johnson, Li & Rabinovic (2007),
   *Adjusting batch effects in microarray expression data using empirical Bayes methods*,
   `Biostatistics <https://doi.org/10.1093/biostatistics/kxj037>`__.

.. [Kang18] Kang *et al.* (2018),
   *Python Implementation of MNN correct*,
   `GitHub <https://github.com/chriscainx/mnnpy>`__.

.. [Korunsky19] Korunsky *et al.* (2019),
   *Fast, sensitive and accurate integration of single-cell data with Harmony*,
   `Nature Methods <https://doi.org/10.1038/s41592-019-0619-0>` __.

.. [Krumsiek10] Krumsiek *et al.* (2010),
   *Odefy – From discrete to continuous models*,
   `BMC Bioinformatics <https://doi.org/10.1186/1471-2105-11-233>`__.

.. [Krumsiek11] Krumsiek *et al.* (2011),
   *Hierarchical Differentiation of Myeloid Progenitors Is Encoded in the Transcription Factor Network*,
   `PLoS ONE <https://doi.org/10.1371/journal.pone.0022649>`__.

.. [Lambiotte09] Lambiotte *et al.* (2009)
   *Laplacian Dynamics and Multiscale Modular Structure in Networks*
   `arXiv <https://arxiv.org/abs/0812.1770>`__.

.. [Lause21] Lause *et al.* (2021)
   *Analytic Pearson residuals for normalization of single-cell RNA-seq UMI data*,
   `Genome Biology <https://doi.org/10.1186/s13059-021-02451-7>`__.

.. [Leek12] Leek *et al.* (2012),
   *sva: Surrogate Variable Analysis. R package*
   `Bioconductor <https://doi.org/10.18129/B9.bioc.sva>`__.

.. [Levine15] Levine *et al.* (2015),
   *Data-Driven Phenotypic Dissection of AML Reveals Progenitor--like Cells that Correlate with Prognosis*,
   `Cell <https://doi.org/10.1016/j.cell.2015.05.047>`__.

.. [Lopez18] Lopez *et al.* (2018),
   *Deep generative modeling for single-cell transcriptomics*
   `Nature Methods <https://doi.org/10.1038/s41592-018-0229-2>`__.

.. [Maaten08] Maaten & Hinton (2008),
   *Visualizing data using t-SNE*,
   `JMLR <http://www.jmlr.org/papers/v9/vandermaaten08a.html>`__.

.. [McCarthy17] McCarthy *et al.* (2017),
   *scater: Single-cell analysis toolkit for gene expression data in R*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btw777>`__.

.. [Moon17] Moon *et al.* (2017),
   *PHATE: A Dimensionality Reduction Method for Visualizing Trajectory Structures in High-Dimensional Biological Data*,
   `BioRxiv <http://biorxiv.org/content/early/2017/03/24/120378>`__.

.. [Satija15] Satija *et al.* (2015),
   *Spatial reconstruction of single-cell gene expression data*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.3192>`__.

.. [Manno18] La Manno *et al.* (2018),
   *RNA velocity of single cells*,
   `Nature <https://doi.org/10.1038/s41586-018-0414-6>`__.

.. [McInnes18] McInnes & Healy (2018),
   *UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction*,
   `arXiv <https://arxiv.org/abs/1802.03426>`__.

.. [Moignard15] Moignard *et al.* (2015),
   *Decoding the regulatory network of early blood development from single-cell gene expression measurements*,
   `Nature Biotechnology <https://doi.org/10.1038/nbt.3154>`__.

.. [Murphy12]
   Murphy (2012,
   *Machine Learning: A Probabilisitc Perspective*,
   `MIT Press <https://mitpress.mit.edu/books/machine-learning-0>`__.

.. [Ntranos18] Ntranos *et al.* (2018),
   *Identification of transcriptional signatures for cell types from single-cell RNA-Seq*,
   `bioRxiv <https://doi.org/10.1101/258566>`__.

.. [Paul15] Paul *et al.* (2015),
   *Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors*,
   `Cell <https://doi.org/10.1016/j.cell.2015.11.013>`__.

.. [Pedersen12] Pedersen (2012),
   *Python implementation of ComBat*
   `GitHub <https://github.com/brentp/combat.py>`__.

.. [Pedregosa11] Pedregosa *et al.* (2011),
   *Scikit-learn: Machine Learning in Python*,
   `JMLR <http://www.jmlr.org/papers/v12/pedregosa11a.html>`__.

.. [Polanski19] Polanski *et al.* (2019),
   *BBKNN: fast batch alignment of single cell transcriptomes*
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btz625>`__.

.. [Plass18] Plass (2018),
   *Cell type atlas and lineage tree of a whole complex animal by single-cell transcriptomics*
   `Science <https://doi.org/10.1126/science.aaq1723>`__.

.. [Scialdone15] Scialdone *et al.* (2015),
   *Computational assignment of cell-cycle stage from single-cell transcriptome data*
   `Methods <https://doi.org/10.1016/j.ymeth.2015.06.021>`__.

.. [Setty16] Setty *et al.* (2016),
   *Wishbone identifies bifurcating developmental trajectories from single-cell data*
   `Nature Biotechnology <https://doi.org/10.1038/nbt.3569>`__.

.. [Setty18] Setty *et al.* (2018),
   *Palantir characterizes cell fate continuities in human hematopoiesis*
   `Nature Biotechnology <https://www.nature.com/articles/s41587-019-0068-4>`__.

.. [Stuart19] Stuart *et al.* (2019),
   *Comprehensive integration of single-cell data*
   `Cell <https://www.sciencedirect.com/science/article/pii/S0092867419305598>`__.

.. [Svensson20] Svensson *et al.* (2020),
   *Interpretable factor models of single-cell RNA-seq via variational autoencoders*
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btaa169>`__.

.. [Nowotschin18i] Nowotschin *et al.* (2018),
   *The emergent landscape of the mouse gut endoderm at single-cell resolution*
   `Nature <https://www.nature.com/articles/s41586-019-1127-1>`__.

.. [Traag17]  (2017),
   *Louvain*,
   `GitHub <https://doi.org/10.5281/zenodo.35117>`__.

.. [Tarashansky19] Tarashansky (2019),
   *Self-assembling manifolds in single-cell RNA sequencing data*,
   `Elife <https://doi.org/10.7554/eLife.48994>`__.

.. [Traag18] Traag *et al.* (2018),
   *From Louvain to Leiden: guaranteeing well-connected communities*
   `arXiv <https://arxiv.org/abs/1810.08473>`__.

.. [Ulyanov16] Ulyanov (2016),
   *Multicore t-SNE*,
   `GitHub <https://github.com/DmitryUlyanov/Multicore-TSNE>`__.

.. [vanDijk18] Van Dijk D *et al.* (2018),
   *Recovering Gene Interactions from Single-Cell Data Using Data Diffusion*,
   `Cell <https://www.cell.com/cell/abstract/S0092-8674(18)30724-4>`__.

.. [Weinreb17] Weinreb *et al.* (2016),
   *SPRING: a kinetic interface for visualizing high dimensional single-cell expression data*,
   `bioRxiv <https://doi.org/10.1101/090332>`__.

.. [Wittmann09] Wittmann *et al.* (2009),
   *Transforming Boolean models to continuous models: methodology and application to T-cell receptor signaling*,
   `BMC Systems Biology <https://doi.org/10.1186/1752-0509-3-98>`__.

.. [Waskom16] Waskom *et al.* (2017),
   *Seaborn*,
   `Zenodo <https://doi.org/10.5281/zenodo.54844>`__.

.. [Wolf18] Wolf *et al.* (2018),
   *Scanpy: large-scale single-cell gene expression data analysis*,
   `Genome Biology <https://doi.org/10.1186/s13059-017-1382-0>`__.

.. [Wolf19] Wolf *et al.* (2019),
   *PAGA: Graph abstraction reconciles clustering with trajectory inference through a topology preserving map of single cells.*
   `Genome Biology <https://doi.org/10.1186/s13059-019-1663-x>`__,
   `bioRxiv <https://doi.org/10.1101/208819>`__.

.. [Wolock19] Wolock *et al.* (2019),
   *BBKNN: fast batch alignment of single cell transcriptomes*
   `Cell Systems  <https://doi.org/10.1016/j.cels.2018.11.005>`__.

.. [Zheng17] Zheng *et al.* (2017),
   *Massively parallel digital transcriptional profiling of single cells*,
   `Nature Communications <https://doi.org/10.1038/ncomms14049>`__.

.. [Zunder15] Zunder *et al.* (2015),
   *A continuous molecular roadmap to iPSC reprogramming through progression analysis of single-cell mass cytometry*,
   `Cell Stem Cell <https://doi.org/10.1016/j.stem.2015.01.015>`__.
